llm_toolkit/code_fixer.py

```diff
@@ -217,123 +217,221 @@ def remove_const_from_png_symbols(content: str) -> str:
-      temp_range[0] = i
-    if temp_range[0] is not None and re.fullmatch(error_end_pattern, line):
-      temp_range[1] = i - 1  # Exclude current line.
-      # In case the original fuzz target was written in C and building with
-      # clang failed, and building with clang++ also failed, we take the
-      # error from clang++, which comes after.
+    # Take the error block that comes later, usually it is the one related to
+    # fuzz target and/or after jcc attempts to fix compile error.
+    if re.fullmatch(clang_diag_error_end_pattern, line):
+      temp_range[1] = i + 1
       error_lines_range = temp_range
       temp_range = [None, None]
 
-  if error_lines_range[0] is not None and error_lines_range[1] is not None:
-    errors.extend(
-        line.rstrip()
-        for line in log_lines[error_lines_range[0]:error_lines_range[1] + 1])
-
-  if not errors:
+  if error_lines_range[0] is None:
     logging.warning('Failed to parse error message from %s.', log_path)
+    return []
+
+  for line in log_lines[error_lines_range[0]:error_lines_range[1]]:
+    errors.append(_strip_color_code(line).rstrip())
   return group_error_messages(errors)
 
 
 def group_error_messages(error_lines: list[str]) -> list[str]:
-  """Groups multi-line error block into one string"""
-  state_unknown = 'UNKNOWN'
-  state_include = 'INCLUDE'
-  state_diag = 'DIAG'
+  """Groups multi-line error block into one string."""
+  clang_pattern = re.compile(r'(\S*:\d+:\d+: .+: .+)|'
+                             r'(In file included from \S*:\d+:)')
+
+  first_line = error_lines[0]
+  if clang_pattern.fullmatch(first_line):
+    return group_clang_errors(error_lines)
+  # Unknown error type, just return lines ungrouped.
+  return error_lines
+
+
+def group_clang_errors(error_lines: list[str]) -> list[str]:
+  """Grouping logics for clang/++ diag errors."""
+  diag_pattern = re.compile(r'(\S*):\d+:\d+: (.+): .+')
+  include_pattern = re.compile(r'In file included from \S*:\d+:')
+  end_pattern = re.compile(r'.*\d+ errors? generated.')
 
-  diag_error_pattern = re.compile(r'(\S*):\d+:\d+: (.+): (.+)')
-  include_error_pattern = re.compile(r'In file included from (\S*):\d+:')
   error_blocks = []
   curr_block = []
-  src_file = ''
-  curr_state = state_unknown
+  include_error_src_file = ''
+  handling_include = False
+  new_block = False
   for line in error_lines:
-    if not line:  # Trim empty lines.
+    if not line:
       continue
 
-    diag_match = diag_error_pattern.fullmatch(line)
-    include_match = include_error_pattern.fullmatch(line)
-
+    diag_match = diag_pattern.match(line)
     if diag_match:
       err_src = diag_match.group(1)
       severity = diag_match.group(2)
 
-      # Matched a note diag line under another diag,
-      # giving help info to fix the previous error.
+      # New error block won't start with note error.
       if severity == 'note':
-        curr_block.append(line)
         continue
```

Comment:
If we do not care about `note`s, could we exclude this case from the pattern above? E.g.,
```python
diag_pattern = re.compile(r'(\S*):\d+:\d+: (?!note)(.+): .+')
```